test/integration/apiserver/field_validation_test.go

```diff
@@ -630,13 +617,13 @@ func testFieldValidationPost(t *testing.T, client clientset.Interface) {
 				SetHeader("Content-Type", tc.contentType).
 				VersionedParams(&tc.opts, metav1.ParameterCodec)
 			result := req.Body(body).Do(context.TODO())
-
-			if result.Error() != nil && len(tc.strictDecodingErrors) == 0 {
-				t.Fatalf("unexpected request err: %v", result.Error())
+			if result.Error() == nil && tc.strictDecodingError != "" {
+				t.Fatalf("received nil error when expecting: %v", tc.strictDecodingError)
 			}
-			for _, strictErr := range tc.strictDecodingErrors {
-				if !strings.Contains(result.Error().Error(), strictErr) {
-					t.Fatalf("missing strict decoding error: %s from error: %v", strictErr, result.Error())
+			if result.Error() != nil && tc.strictDecodingError != result.Error().Error() {
+				if result.Error() != nil && !strings.HasSuffix(result.Error().Error(), tc.strictDecodingError) {
```

Comment:
I think success is defined as:
There is no error **AND** you're not expecting one **OR**
There is an error **AND** the error is suffixed with the expected error.

Since we're trying to find the failure cases, applying De Morgan's a few times gives us:
there is an error **OR** we expect one **AND**
there is no error **OR** the error is not suffixed with the expected error.

We can develop that expression into (and re-ordering a little):
there is an error **AND** there is no error (that's always false, we can drop it) **OR**
you're expecting an error **AND** there is no error (that's line 620) **OR**
there is an error **AND** the error is not suffixed with the expected error **OR**
you're expecting an error **AND** the error is not suffixed with the expected error

We can factorize these last two conditions into one probably:
(there is an error **AND** you're expecting an error) **OR** the error is not suffixed with the expected error

I think that last statement should probably be written as:
```
if result.Error() != nil && tc.strictDecodingError != "" && !strings.HasSuffix(result.Error().Error(), tc.strictDecodingError)
```

Don't you have a bug here where `tc.strictDecodingError` is empty string will succeed? Is my math so terrible? Is this the worst comment in this history of comments?

Reply:
This seems to apply in other places

Reply:
If we are okay giving the same error message when across errors 2 and 3, I think we can simplify further to:

```
if result.Error() == nil && tc.strictDecodingError != "" {
        t.Fatalf("received nil error when expecting: %v", tc.strictDecodingError)
}
if result.Error() != nil && (tc.strictDecodingError == "" || !strings.HasSuffix(result.Error().Error(), tc.strictDecodingError)) {
        t.Fatalf("expected error: %s, got: %v", tc.strictDecodingError, result.Error())
}
```

Reply:
Yeah, that's good, maybe quote the `strictDecodingError` for when it's empty?
```
t.Fatalf("expected error: %q, got: %v", tc.strictDecodingError, result.Error())
```

Reply:
LGTM after that

Reply:
done for both clauses